### PR TITLE
(feat) Add possibility to transfer entire balance.

### PR DESCRIPTION
**Pull request by bznein**
## Description

closes: #6567

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

**Review by crodriguezvega — COMMENTED** [review]
Changes look good to me, @bznein. Maybe it would be worth adding a short e2e test and check that the balance of the account is zero after the transfer. Might be worth using 2 denoms in the e2e test.

**Comment by bznein** [comment]
> Changes look good to me, @bznein. Maybe it would be worth adding a short e2e test and check that the balance of the account is zero after the transfer. Might be worth using 2 denoms in the e2e test.

Thanks Carlos! I've added the test, LMK if this is what you meant. FIY the test failed locally unless I manually built a new image and used that. I assume this would not happen in the CI.

File: modules/apps/transfer/keeper/relay.go [hunk 1]
```diff
@@ -98,6 +98,12 @@ func (k Keeper) sendTransfer(
 	tokens := make([]types.Token, 0, len(coins))
 
 	for _, coin := range coins {
+
+		// Using types.UnboundedSpendLimit allows us to send the entire balance of a given denom.
+		if coin.Amount.Equal(types.UnboundedSpendLimit()) {
```
[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/keeper/relay.go:102]: Agreed :) I was a bit surprised to see it there, we can definitely move it somewhere else

**Review by gjermundgaraba — APPROVED** [review]
Nice add-on! LGTM 🚀

File: modules/apps/transfer/keeper/relay.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/apps/transfer/keeper/relay.go:102]: yeah, sounds good to move it.

[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/keeper/relay.go:102]: Done!

File: e2e/tests/transfer/base_test.go [hunk 2]
```diff
@@ -553,3 +553,77 @@ func (s *TransferTestSuite) TestMsgTransfer_WithMemo() {
...
+		s.Require().Equal(coinFromA.Amount.Int64(), actualBalance.Int64())
+
+		actualBalance, err = query.Balance(ctx, chainA, chainAAddress, chainADenom)
+
+		s.Require().NoError(err)
+		s.Require().Equal(testvalues.StartingTokenAmount-coinFromA.Amount.Int64(), actualBalance.Int64())
+	})
+
+	t.Run("send entire balance from B to A", func(t *testing.T) {
+		transferTxResp := s.Transfer(ctx, chainB, chainBWallet, channelA.Counterparty.PortID, channelA.Counterparty.ChannelID, sdk.NewCoins(sdk.NewCoin(chainBIBCToken.IBCDenom(), transfertypes.UnboundedSpendLimit())), chainBAddress, chainAAddress, s.GetTimeoutHeight(ctx, chainB), 0, "", nil)
```
Comment (crodriguezvega) [inline e2e/tests/transfer/base_test.go:613]: Should we transfer the whole balance for both the native chain B denom and the IBC denom? Just to test that it all works fine with multidenom.

**Review by crodriguezvega — APPROVED** [review]
Thanks, @bznein. I left just one comment.

File: e2e/tests/transfer/base_test.go (hunk 2 above)
Comment (damiannolan) [reply e2e/tests/transfer/base_test.go:613]: Why not!

File: CHANGELOG.md [hunk 3]
```diff
@@ -82,6 +82,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 * (apps/transfer) [\#6492](https://github.com/cosmos/ibc-go/pull/6492) Added new `Tokens` field to `MsgTransfer` to enable sending of multiple denoms, and deprecated the `Token` field.
 * (apps/transfer) [\#6693](https://github.com/cosmos/ibc-go/pull/6693) Added new `Forwarding` field to `MsgTransfer` to enable forwarding tokens through multiple intermediary chains with a single transaction. This also enables automatic unwinding of tokens to their native chain. `x/authz` support for transfer allows granters to specify a set of possible forwarding hops that are allowed for grantees.
+* (apps/transfer) [\#6877](https://github.com/cosmos/ibc-go/pull/6877) Added the possibility to transfer the entire balance of a particular denomination by using as amount a value equal to [`UnboundedSpendLimit`](https://github.com/cosmos/ibc-go/blob/715f00eef8727da41db25fdd4763b709bdbba07e/modules/apps/transfer/types/transfer_authorization.go#L253-L255)
```
Comment (damiannolan) [inline CHANGELOG.md:85]: mega nit, I think this changelog reads a little odd, maybe we could adjust it slightly

```suggestion
* (apps/transfer) [\#6877](https://github.com/cosmos/ibc-go/pull/6877) Added the possibility to transfer the entire user balance of a particular denomination by using [`UnboundedSpendLimit`](https://github.com/cosmos/ibc-go/blob/715f00eef8727da41db25fdd4763b709bdbba07e/modules/apps/transfer/types/transfer_authorization.go#L253-L255) as the token amount.
```

**Review by damiannolan — APPROVED** [review]
Nice work, lgtm!

**Comment by crodriguezvega** [comment]
@bznein Before merging the PR, should we address the latest comments?

File: e2e/tests/transfer/base_test.go (hunk 2 above)
Comment (bznein) [reply e2e/tests/transfer/base_test.go:613]: Good idea, done

File: modules/apps/transfer/types/token.go [hunk 4]
```diff
@@ -42,3 +42,12 @@ func (t Token) ToCoin() (sdk.Coin, error) {
...
 }
+
+// UnboundedSpendLimit returns the sentinel value that can be used
+// as the amount for a denomination's spend limit for which spend limit updating
+// should be disabled. Please note that using this sentinel value means that a grantee
+// will be granted the privilege to do ICS20 token transfers for the total amount
+// of the denomination available at the granter's account.
+func UnboundedSpendLimit() sdkmath.Int {
+	return sdkmath.NewIntFromBigInt(maxUint256)
+}
```
[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/types/token.go:53]: > Maybe its just me that needs a little bit of adjusting to using this in the flow of regular transfer token sends but the `SpendLimit` terminology was something added specifically with the authz functionality, so it makes me think of that.


Yeah I do see the issue there. This is "new" to me so I didn't mind moving it. 

> I would recommend also moving the `maxUint256` var over to this file.

Done!

File: CHANGELOG.md (hunk 3 above)
Comment (bznein) [reply CHANGELOG.md:85]: Done
